File: apps/macos/PasteSheets/Data/DataSources/PasteItemDataSource.swift
import Foundation
import SQLite3

protocol PasteItemDataSource {
    func fetchAll() throws -> [PasteItemDTO]
    func insert(content: String, directory: String, memo: String?) throws -> Int64
    func update(id: Int64, content: String, directory: String, memo: String?) throws
    func delete(id: Int64) throws
    func findByContent(_ content: String, directory: String) throws -> PasteItemDTO?
    func countByDirectory(_ directory: String) throws -> Int64
    func deleteOldest(directory: String, excess: Int64) throws
}

final class PasteItemDataSourceImpl: PasteItemDataSource {
    private let db = DatabaseManager.shared

    func fetchAll() throws -> [PasteItemDTO] {
        try db.query(
            "SELECT id, content, directory, created_at, memo FROM paste_sheets ORDER BY created_at DESC",
            mapper: Self.mapRow
        )
    }

    func insert(content: String, directory: String, memo: String?) throws -> Int64 {
        try db.executeReturningId(
            "INSERT INTO paste_sheets (content, directory, memo) VALUES (?1, ?2, ?3)",
            params: [content, directory, memo]
        )
    }

    func update(id: Int64, content: String, directory: String, memo: String?) throws {
        try db.execute(
            "UPDATE paste_sheets SET content = ?1, directory = ?2, memo = ?3, created_at = CURRENT_TIMESTAMP WHERE id = ?4",
            params: [content, directory, memo, id]
        )
    }

    func delete(id: Int64) throws {
        try db.execute("DELETE FROM paste_sheets WHERE id = ?1", params: [id])
    }

    func findByContent(_ content: String, directory: String) throws -> PasteItemDTO? {
        try db.queryOne(
            "SELECT id, content, directory, created_at, memo FROM paste_sheets WHERE content = ?1 AND directory = ?2 LIMIT 1",
            params: [content, directory],
            mapper: Self.mapRow
        )
    }

    func countByDirectory(_ directory: String) throws -> Int64 {
        let result = try db.queryOne(
            "SELECT COUNT(*) FROM paste_sheets WHERE directory = ?1",
            params: [directory]
        ) { stmt in
            sqlite3_column_int64(stmt, 0)
        }
        return result ?? 0
    }

    func deleteOldest(directory: String, excess: Int64) throws {
        try db.execute(
            """
            DELETE FROM paste_sheets WHERE id IN (
                SELECT id FROM paste_sheets
                WHERE directory = ?1
                ORDER BY created_at ASC
                LIMIT ?2
            )
            """,
            params: [directory, excess]
        )
    }

    private static func mapRow(_ stmt: OpaquePointer) -> PasteItemDTO {
        PasteItemDTO(
            id: sqlite3_column_int64(stmt, 0),
            content: String(cString: sqlite3_column_text(stmt, 1)),
            directory: String(cString: sqlite3_column_text(stmt, 2)),
            createdAt: String(cString: sqlite3_column_text(stmt, 3)),
            memo: sqlite3_column_text(stmt, 4).map { String(cString: $0) }
        )
    }
}

File: apps/windows/PasteSheet/Data/DataSources/PasteItemDataSource.cs
using Microsoft.Data.Sqlite;
using PasteSheet.Data.Database;
using PasteSheet.Data.DTOs;

namespace PasteSheet.Data.DataSources;

public interface IPasteItemDataSource
{
    List<PasteItemDTO> FetchAll();
    long Insert(string content, string directory, string? memo);
    void Update(long id, string content, string directory, string? memo);
    void Delete(long id);
    PasteItemDTO? FindByContent(string content, string directory);
    long CountByDirectory(string directory);
    void DeleteOldest(string directory, long excess);
}

public sealed class PasteItemDataSource : IPasteItemDataSource
{
    private readonly DatabaseManager _db = DatabaseManager.Shared;

    public List<PasteItemDTO> FetchAll() =>
        _db.Query(
            "SELECT id, content, directory, created_at, memo FROM paste_sheets ORDER BY created_at DESC",
            null, MapRow);

    public long Insert(string content, string directory, string? memo) =>
        _db.ExecuteReturningId(
            "INSERT INTO paste_sheets (content, directory, memo) VALUES (@p1, @p2, @p3)",
            new object?[] { content, directory, memo });

    public void Update(long id, string content, string directory, string? memo) =>
        _db.Execute(
            "UPDATE paste_sheets SET content = @p1, directory = @p2, memo = @p3, created_at = CURRENT_TIMESTAMP WHERE id = @p4",
            new object?[] { content, directory, memo, id });

    public void Delete(long id) =>
        _db.Execute("DELETE FROM paste_sheets WHERE id = @p1", new object?[] { id });

    public PasteItemDTO? FindByContent(string content, string directory) =>
        _db.QueryOne(
            "SELECT id, content, directory, created_at, memo FROM paste_sheets WHERE content = @p1 AND directory = @p2 LIMIT 1",
            new object?[] { content, directory }, MapRow);

    public long CountByDirectory(string directory) =>
        _db.QueryOne(
            "SELECT COUNT(*) FROM paste_sheets WHERE directory = @p1",
            new object?[] { directory },
            r => r.GetInt64(0));

    public void DeleteOldest(string directory, long excess) =>
        _db.Execute(
            """
            DELETE FROM paste_sheets WHERE id IN (
                SELECT id FROM paste_sheets
                WHERE directory = @p1
                ORDER BY created_at ASC
                LIMIT @p2
            )
            """,
            new object?[] { directory, excess });

    private static PasteItemDTO MapRow(SqliteDataReader r) =>
        new(
            r.GetInt64(0),
            r.GetString(1),
            r.GetString(2),
            r.GetString(3),
            r.IsDBNull(4) ? null : r.GetString(4));
}
